defi/order-book-dex/src/trading_pair.rs

```diff
@@ -0,0 +1,311 @@
+use std::cmp::min;
+
+use scrypto::prelude::*;
+
+use crate::model::*;
+
+blueprint! {
+
+    /// This component represents a trading pair like e.g. XRD/rUSD. It can be used by traders to
+    /// create limit and market orders.
+    struct TradingPair {
+        /// Internal authority for minting order NFRs (non fungible resources)
+        minter: Vault,
+
+        /// A NFR that represents limit orders
+        order_resource: ResourceDef,
+
+        /// The order book that holds all limit orders that are created for this trading pair
+        order_book: OrderBook,
+
+        /// The resources that market makers have deposited into this component by creating limit ask orders
+        base_funds: Vault,
+
+        /// The resources that market makers have deposited into this component by creating limit bid orders
+        quote_funds: Vault,
+    }
+
+    impl TradingPair {
+        /// Instantiates a new TradingPair component for the given base_resource and quote_resource.
+        /// To create a trading pair XRD/rUSD, one would issue a call like this: `instantiate(xrd_address, rusd_address)`
+        pub fn instantiate(base_resource: ResourceDef, quote_resource: ResourceDef) -> Component {
+            assert_is_fungible(&base_resource);
+            assert_is_fungible(&quote_resource);
+
+            let minter =
+                ResourceBuilder::new_fungible(DIVISIBILITY_NONE).initial_supply_fungible(1);
+
+            // This NFR represents the orders managed by this trading pair.
+            // Orders an be minted, burned and updated by this component.
+            let order_resource = ResourceBuilder::new_non_fungible()
+                .metadata(
+                    "name",
+                    format!(
+                        "Dex limit order ({}/{})",
+                        base_resource.address(),
+                        quote_resource.address()
+                    ),
+                )
+                .flags(MINTABLE | BURNABLE | INDIVIDUAL_METADATA_MUTABLE)
+                .badge(minter.resource_def(), ALL_PERMISSIONS)
```

Comment:
It's ok right now to put `ALL_PERMISSIONS`, but you wouldn't want that for production because it would mean that the minter authority would be able to update the flags (MAY_MANAGE_RESOURCE_FLAGS). It's not a big deal since the badge is stored on the component and you don't provide methods to change the resource flags, but we are planning on having the wallet display warnings when a token might be recallable or restricted in the future

Reply:
Hmm, I totally missed the MAY_MANAGE_RESOURCE_FLAGS flag and thought the ALL_PERMISSIONS flag only pertained to the set flags (MINTABLE, BURNABLE, INDIVIDUAL_METADATA_MUTABLE). I will correct that and only use the necessary permissions.

Now I'm a bit sceptical about the ALL_FLAGS and ALL_PERMISSIONS constants in general. In security centric code like blueprints, would we not be well adviced to only very selectively set flags and permissions? Usage of these constants should be discouraged imho.

Reply:
I'll ask to the team about that